File: Android/Hybrid/src/main/java/siminov/hybrid/model/LibraryDescriptor.java
/**
 * [SIMINOV FRAMEWORK - HYBRID]
 * Copyright [2014-2016] [Siminov Software Solution LLP|support@siminov.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/


package siminov.hybrid.model;

import java.util.Collection;
import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;


/**
 * Exposes methods to GET and SET Library Descriptor information as per define in LibraryDescriptor.xml file by application.
	<p>
		<pre>
		
Example:
	{@code
	
		
	<!-- DESIGN OF LibraryDescriptor.xml -->
	
	<library-descriptor>
	
	    <!-- General Properties Of Library -->
	    
	    <!-- Mandatory Field -->
		<property name="name">name_of_library</property>
		
		<!-- Optional Field -->
		<property name="description">description_of_library</property>
	
		
		
		<!-- Entity Descriptor Needed Under This Library Descriptor -->
		
		<!-- Optional Field -->
			<!-- Entity Descriptors -->
		<entity-descriptors>
			<entity-descriptor>name_of_database_descriptor.full_path_of_entity_descriptor_file</entity-descriptor>
		</entity-descriptors>
		 
		
		<!-- Service Descriptors -->
			
		<!-- Optional Field -->
			<!-- Service Descriptor -->
		<service-descriptors>
		    <service-descriptor>full_path_of_service-descriptor_file</service-descriptor>
		</service-descriptors>
		
		
		<!-- Sync Descriptors -->
		
		<!-- Optional Field -->
			<!-- Sync Descriptor -->
		<sync-descriptors>
		    <sync-descriptor>full_path_of_sync_descriptor_file</sync-descriptor>
		</sync-descriptors>
		
		
		<!-- Adapter Descriptors -->
		
		<!-- Optional Field -->
			<!-- Adapter Descriptor -->
		<adapter-descriptors>
		    <adapter-descriptor>full_path_of_adapter_descriptor_file</adapter-descriptor>
		</adapter-descriptors>	
			
		
	</library-descriptor>
	
	

	
		</pre>
	</p>
 *
 */
public class LibraryDescriptor extends siminov.connect.model.LibraryDescriptor {

	private Collection<String> adapterDescriptorPaths = new ConcurrentLinkedQueue<String> ();

	private Map<String, AdapterDescriptor> adapterDescriptorsBasedOnPath = new ConcurrentHashMap<String, AdapterDescriptor>();
	private Map<String, AdapterDescriptor> adapterDescriptorsBasedOnName = new ConcurrentHashMap<String, AdapterDescriptor>();
	
	/**
	 * Get All Adapter Descriptor Paths.
	 * @return All Adapter Descriptor Paths.
	 */
	public Iterator<String> getAdapterDescriptorPaths() {
		return this.adapterDescriptorPaths.iterator();
	}
	
	/**
	 * Get All Adapter Descriptors.
	 * @return All Adapter Descriptors.
	 */
	public Iterator<AdapterDescriptor> getAdapterDescriptors() {
		return this.adapterDescriptorsBasedOnName.values().iterator();
	}
	
	/**
	 * Get Adapter Descriptor Based on adapter descriptor name.
	 * @param adapterDescriptorName Name of Adapter Descriptor.
	 * @return Adapter Descriptor.
	 */
	public AdapterDescriptor getAdapterDescriptorBasedOnName(String adapterDescriptorName) {
		return this.adapterDescriptorsBasedOnName.get(adapterDescriptorName);
	}
	
	/**
	 * Get Adapter Descriptor Based on adapter descriptor path.
	 * @param adapterDescriptorPath Path of Adapter Descriptor.
	 * @return Adapter Descriptor.
	 */
	public AdapterDescriptor getAdapterDescriptorBasedOnPath(String adapterDescriptorPath) {
		return this.adapterDescriptorsBasedOnPath.get(adapterDescriptorPath);
	}
	
	/**
	 * Add Adapter Descriptor Path.
	 * @param adapterDescriptorPath Path of Adapter Descriptor.
	 */
	public void addAdapterDescriptorPath(String adapterDescriptorPath) {
		this.adapterDescriptorPaths.add(adapterDescriptorPath);
	}

	/**
	 * Add Adapter Descriptor based on adapter descriptor path.
	 * @param adapterDescriptorPath Path of Adapter Descriptor.
	 * @param adapterDescriptor Adapter Descriptor.
	 */
	public void addAdapterDescriptor(String adapterDescriptorPath, AdapterDescriptor adapterDescriptor) {
		this.adapterDescriptorsBasedOnPath.put(adapterDescriptorPath, adapterDescriptor);
		this.adapterDescriptorsBasedOnName.put(adapterDescriptor.getName(), adapterDescriptor);
	}
	
	/**
	 * Check whether Adapter Descriptor exist based on adapter descriptor path.
	 * @param adapterDescriptorPath Path of Adapter Descriptor.
	 * @return true/false; TRUE if adapter descriptor exist, FALSE if adapter descriptor does not exist.
	 */
	public boolean containAdapterDescriptorBasedOnPath(String adapterDescriptorPath) {
		return this.adapterDescriptorsBasedOnPath.containsKey(adapterDescriptorPath);
	}
	
	/**
	 * Check whether Adapter Descriptor exist based on adapter name.
	 * @param adapterDescriptorName Name of Adapter Descriptor.
	 * @return true/false; TRUE if adapter descriptor exist, FALSE if adapter descriptor does not exist.
	 */
	public boolean containAdapterDescriptorBasedOnName(String adapterDescriptorName) {
		return this.adapterDescriptorsBasedOnName.containsKey(adapterDescriptorName);
	}
	
	/**
	 * Remove Adapter Descriptor based on adapter descriptor name.
	 * @param adapterDescriptorName Name of Adapter Descriptor.
	 */
	public void removeAdapterDescriptorBasedOnName(String adapterDescriptorName) {
		Iterator<String> libraryPaths = this.adapterDescriptorsBasedOnPath.keySet().iterator();
		
		boolean found = false;
		String keyMatched = null;
		while(libraryPaths.hasNext()) {
			String libraryPath = libraryPaths.next();
			
			AdapterDescriptor adapterDescriptor = this.adapterDescriptorsBasedOnPath.get(libraryPath);
			if(adapterDescriptor.getName().equalsIgnoreCase(adapterDescriptorName)) {
				keyMatched = libraryPath;
				found = true;
				break;
			}
		}
		
		if(found) {
			removeAdapterBasedOnPath(keyMatched);
		}
	}
	
	/**
	 * Remove Adapter Descriptor based on adapter descriptor path.
	 * @param adapterPath Path of Adapter Descriptor.
	 */
	public void removeAdapterBasedOnPath(String adapterPath) {
		AdapterDescriptor adapaterDescriptor = this.adapterDescriptorsBasedOnPath.get(adapterPath);

		this.adapterDescriptorsBasedOnName.remove(adapaterDescriptor.getName());
		this.adapterDescriptorsBasedOnPath.remove(adapterPath);
	}
	
	/**
	 * Remove Adapter Descriptor.
	 * @param adapterDescriptor Adapter Descriptor.
	 */
	public void removeAdapater(AdapterDescriptor adapterDescriptor) {
		removeAdapterDescriptorBasedOnName(adapterDescriptor.getName());
	}
	
}

File: JavaScript/Siminov/Model/LibraryDescriptor.js
/**
 * [SIMINOV FRAMEWORK - HYBRID]
 * Copyright [2014-2016] [Siminov Software Solution LLP|support@siminov.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/




/**
	It contain all Models as per required by Siminov Framework.

	@module Model
*/

var win;
var dom;

try {

    if(!window) {
    	window = global || window;
    }

	win = window;
	dom = window['document'];
} catch(e) {
	win = Ti.App.Properties;
}


if(dom == undefined) {
    var Constants = require('../Constants');
    var Dictionary = require('../Collection/Dictionary');
    
    
    module.exports = LibraryDescriptor;
    win.LibraryDescriptor = LibraryDescriptor;
}

/**
 	Exposes methods to GET and SET Library Descriptor information as per define in LibraryDescriptor.xml file by application.
		
	Example:

		<!-- DESIGN OF LibraryDescriptor.xml -->

		<library-descriptor>

			<!-- General Properties Of Library -->

			<!-- Mandatory Field -->
			<property name="name">name_of_library</property>

			<!-- Optional Field -->
			<property name="description">description_of_library</property>



			<!-- Entity Descriptor Needed Under This Library Descriptor -->

			<!-- Optional Field -->
				<!-- Database Mappings -->
			<entity-descriptors>
				<entity-descriptor>name_of_database_descriptor.full_path_of_database_mapping_descriptor_file</entity-descriptor>
			</entity-descriptors>


			<!-- Service Descriptors -->

			<!-- Optional Field -->
				<!-- Service Descriptor -->
			<service-descriptors>
				<service-descriptor>full_path_of_service-descriptor_file</service-descriptor>
			</service-descriptors>


			<!-- Sync Descriptors -->

			<!-- Optional Field -->
				<!-- Sync Descriptor -->
			<sync-descriptors>
				<sync-descriptor>full_path_of_sync_descriptor_file</sync-descriptor>
			</sync-descriptors>


			<!-- Adapter Descriptors -->

			<!-- Optional Field -->
				<!-- Adapter Descriptor -->
			<adapter-descriptors>
				<adapter-descriptor>full_path_of_adapter_descriptor_file</adapter-descriptor>
			</adapter-descriptors>


		</library-descriptor>
	
	@module Model		
	@class LibraryDescriptor
	@constructor
 */
function LibraryDescriptor() {

    var properties = new Dictionary();

    var entityDescriptorPaths = new Array();
	var serviceDescriptorPaths = new Array();
	var adapterDescriptorPaths = new Array();
	


	/**
	 	Get library name.
	 	
	 	@method getName
	 	@return {String} Name of Library
	 */
    this.getName = function() {
    	return properties.get(Constants.LIBRARY_DESCRIPTOR_NAME);
	}
	
	/**
	 	Set library name as per defined in LibraryDescriptor.xml
	 
	 	@method setName
 		@param name {String} Name of Library
	 */
    this.setName = function(name) {
    	properties.add(Constants.LIBRARY_DESCRIPTOR_NAME, name);
	}
	
	/**
	 	Get descriptor as per defined in LibraryDescriptor.xml
	 	
	 	@method getDescription
	 	@return {String} Description of Library
	 */
    this.getDescription = function() {
    	return properties.get(Constants.LIBRARY_DESCRIPTOR_DESCRIPTION);
	}
	
	/**
	 	Set description as per defined in LibraryDescritor.core.xml
	 
	 	@method setDescription
	 	@param descriptor {String} Description of Library
	 */
    this.setDescription = function(description) {
    	properties.add(Constants.LIBRARY_DESCRIPTOR_DESCRIPTION, description);
	}
	
		
	/**
	 	Get all Properties defined in descriptor.
	 
	 	@method getProperties
		@return {Array} All Property Values.
	 */
	this.getProperties = function() {
		return properties.values();
	}


	/**
	 	Get Property based on name provided.
	 
	 	@method getProperty
		@param name {String} Name of Property.
		@return {String} Property value.
	 */
	this.getProperty = function(name) {
		return properties.get(name);
	}


	/**
	 	Check whether Property exist or not.
	 
	 	@method containProperty
		@param name {String} Name of Property.
	 	@return {Boolean} true/false, TRUE if property exist, FALSE if property does not exist.
	 */
	this.containProperty = function(name) {
		return properties.exists(name);
	}


	/**
	 	Add Property in property pool.
	 
	 	@method addProperty
		@param name {String} Name of Property.
	 	@param value {String} value of Property.
	 */
	this.addProperty = function(name, value) {
		properties.add(name, value);
	}
	

	/**
	 	Remove Property from property pool.
	 
	 	@method  removeProperty
		@param name {String} Name of Property.
	 */
	this.removeProperty = function(name) {
		properties.remove(name);
	}

	
	/**
	 	Get all entity descriptor paths as per defined in DatabaseDescriptor.xml file.
	 
	 	@method getEntityDescriptorPaths
	 	@return {Array} It contain all entity paths.
	 */
    this.getEntityDescriptorPaths = function() {
    	return entityDescriptorPaths;
	}
	
	/**
	 	Add entity path as per defined in DatabaseDescriptor.xml file.
	 		
		EXAMPLE:
			<database-descriptor>
				<entity-descriptors>
					<entity-descriptor>Entity-Descriptors/Book.xml</entity-descriptor>
					<entity-descriptor>Entity-Descriptors/Lession.xml</entity-descriptor>
				</entity-descriptors>
			</database-descriptor>
	
	 
	 	@method addEntityDescriptorPath
	 	@param entityDescriptorPath {String} Entity Descriptor Path.
	 */
    this.addEntityDescriptorDescriptorPath = function(entityDescriptorPath) {
    	entityDescriptorPaths.push(entityDescriptorPaths);
	}
    
    
    /**
     * Get all service descriptor paths
     * 
     * @method getServiceDescriptorPaths
     * @return {Array} Service Descriptor Paths
     */
    this.getServiceDescriptorPaths = function() {
    	return serviceDescriptorPaths;
    }
    
    /**
     * Add service descriptor path
     * 
     * @method addServiceDescriptorPath
     * @param serviceDescriptorPath {String} Service descriptor path
     */
    this.addServiceDescriptorPath = function(serviceDescriptorPath) {
    	serviceDescriptorPaths.push(serviceDescriptorPath);
    }
    
    /**
     * Get all adapter descriptor paths
     * 
     * @method getAdapterDescriptorPaths
     * @param {Array} Adapter Descriptor Paths
     */
    this.getAdapterDescriptorPaths = function() {
    	return adapterDescriptorPaths;
    }
    
    /**
     * Get adapter descriptor path
     * 
     * @method getAdapterDescriptorPath
     * @param adapterDescriptorPath {String} Adapter Descriptor Path
     */
    this.getAdapterDescriptorPath = function(adapterDescriptorPath) {
    	adapterDescriptorPaths.push(adapterDescriptorPath);
    }
    
}
